overlord/devicestate/handlers_systems.go

```diff
@@ -121,6 +132,154 @@ func purgeNewSystemSnapFiles(logfile string) error {
+
+	if err := targetSeed.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+		return nil, err
+	}
+
+	var uniqueToTarget []string
+	err = targetSeed.Iter(func(sn *seed.Snap) error {
+		if sn.ID() != "" && !requiredByOtherSystems[sn.Path] {
+			uniqueToTarget = append(uniqueToTarget, sn.Path)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return uniqueToTarget, nil
+}
+
+func (m *DeviceManager) doRemoveRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	systems, err := m.systems()
+	if err != nil {
+		return fmt.Errorf("cannot get recovery systems: %w", err)
+	}
+
+	setup, err := taskRemoveRecoverySystemSetup(t)
+	if err != nil {
+		return err
+	}
+
+	found := false
+	for _, sys := range systems {
+		if sys.Label == setup.Label {
+			found = true
+			if sys.Current {
+				return fmt.Errorf("cannot remove current recovery system: %q", setup.Label)
+			}
+			break
+		}
+	}
+
+	// if we couldn't find the system in the returned list of systems, then we
+	// might have already attempted to remove it. in that case, we should do our
+	// best effort at making sure that it is fully removed.
+
+	if found && len(systems) == 1 {
+		return fmt.Errorf("cannot remove last recovery system: %q", setup.Label)
+	}
+
+	deviceCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return fmt.Errorf("cannot get device context: %w", err)
+	}
+
+	// if this task has partially run before, then we might have already removed
+	// some files required to load the being-removed system seed. to avoid this,
+	// we first check if we already have stored a list of snaps to remove
+	// (meaning, this task is being re-run). if the list isn't present, then we
+	// calculate it and store it in the task state for potential future re-runs.
+	var snapsToRemove uniqueSnapsInRecoverySystem
+	if err := t.Get("snaps-to-remove", &snapsToRemove); err != nil {
+		if !errors.Is(err, state.ErrNoState) {
+			return fmt.Errorf("cannot get snaps to remove from task: %w", err)
+		}
+
+		uniqueSnapPaths, err := snapsUniqueToRecoverySystem(setup.Label, systems)
+		if err != nil {
+			return fmt.Errorf("cannot get snaps unique to recovery system %q: %w", setup.Label, err)
+		}
+
+		snapsToRemove.SnapPaths = uniqueSnapPaths
+
+		t.Set("snaps-to-remove", snapsToRemove)
+	}
+
+	recoverySystemsDir := filepath.Join(boot.InitramfsUbuntuSeedDir, "systems")
+
+	if err := boot.DropRecoverySystem(deviceCtx, setup.Label); err != nil {
+		return fmt.Errorf("cannot drop recovery system %q: %v", setup.Label, err)
+	}
+
+	for _, sn := range snapsToRemove.SnapPaths {
+		path := filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps", filepath.Base(sn))
+		if err := os.RemoveAll(path); err != nil {
+			return fmt.Errorf("cannot remove snap %q: %w", path, err)
+		}
+	}
+
+	if err := os.RemoveAll(filepath.Join(recoverySystemsDir, setup.Label)); err != nil {
+		return fmt.Errorf("cannot remove recovery system %q: %w", setup.Label, err)
+	}
+
```

Comment:
to avoid wasted effort we want to set t.SetStatus(DoneStatus) here?  the issue is that we will release the lock before this is done by the framework

Reply:
Will do. Is there ever a case where you shouldn't mark a task as done at the end of the handler?

Reply:
not really, this is a bit of quirk of the framework. In retrospect we should have started tasks with the lock taken. I have some ideas to fix this at some point but it's hard to prioritize